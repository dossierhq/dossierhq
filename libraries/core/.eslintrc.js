module.exports = {
  env: {
    browser: true,
    es2021: true,
  },
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:import/recommended',
    'plugin:import/typescript',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 12,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint'],
  reportUnusedDisableDirectives: true,
  rules: {
    '@typescript-eslint/consistent-type-imports': ['warn'],
    '@typescript-eslint/no-parameter-properties': ['error'],
    '@typescript-eslint/no-unused-vars': [
      'warn',
      {
        vars: 'local',
        args: 'after-used',
        ignoreRestSiblings: true,
        varsIgnorePattern: '^_',
        argsIgnorePattern: '^_',
      },
    ],
    //TODO figure out how to deal with imports. jest struggles with .js
    // 'import/extensions': ['error', { js: 'always', jsx: 'always', ts: 'always', tsx: 'always' }],
    // 'import/no-unresolved': ['off'],
  },
};
