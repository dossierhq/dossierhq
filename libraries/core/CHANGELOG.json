{
  "name": "@dossierhq/core",
  "entries": [
    {
      "version": "0.2.19",
      "tag": "@dossierhq/core_v0.2.19",
      "date": "Tue, 13 Jun 2023 21:42:08 GMT",
      "comments": {}
    },
    {
      "version": "0.2.18",
      "tag": "@dossierhq/core_v0.2.18",
      "date": "Tue, 13 Jun 2023 20:53:31 GMT",
      "comments": {
        "none": [
          {
            "comment": "Enable schemas where adminOnly fields refer to adminOnly types"
          }
        ]
      }
    },
    {
      "version": "0.2.17",
      "tag": "@dossierhq/core_v0.2.17",
      "date": "Tue, 13 Jun 2023 15:58:39 GMT",
      "comments": {
        "none": [
          {
            "comment": "Normalize schema. Prevent changing adminOnly of types. Prevent changing field types, removing fields in schema. Prevent duplicate field names in schema. Enable partial update of schema. Validate integer on number fields. Break: Replace AdminSchema.mergeWith() with .updateAndValidate()."
          }
        ]
      }
    },
    {
      "version": "0.2.16",
      "tag": "@dossierhq/core_v0.2.16",
      "date": "Tue, 09 May 2023 22:47:11 GMT",
      "comments": {}
    },
    {
      "version": "0.2.15",
      "tag": "@dossierhq/core_v0.2.15",
      "date": "Tue, 09 May 2023 22:20:37 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add values to string field schema"
          }
        ]
      }
    },
    {
      "version": "0.2.14",
      "tag": "@dossierhq/core_v0.2.14",
      "date": "Sun, 23 Apr 2023 21:31:27 GMT",
      "comments": {
        "none": [
          {
            "comment": "Filter admin search/sample on validity"
          }
        ]
      }
    },
    {
      "version": "0.2.13",
      "tag": "@dossierhq/core_v0.2.13",
      "date": "Tue, 11 Apr 2023 19:08:01 GMT",
      "comments": {}
    },
    {
      "version": "0.2.12",
      "tag": "@dossierhq/core_v0.2.12",
      "date": "Tue, 11 Apr 2023 18:39:14 GMT",
      "comments": {}
    },
    {
      "version": "0.2.11",
      "tag": "@dossierhq/core_v0.2.11",
      "date": "Tue, 11 Apr 2023 15:53:50 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add valid to AdminEntityStatus. Add validateEntityInfo()."
          }
        ]
      }
    },
    {
      "version": "0.2.10",
      "tag": "@dossierhq/core_v0.2.10",
      "date": "Tue, 04 Apr 2023 09:26:20 GMT",
      "comments": {
        "none": [
          {
            "comment": "Move entity info validation to core and validate name and version"
          }
        ]
      }
    },
    {
      "version": "0.2.9",
      "tag": "@dossierhq/core_v0.2.9",
      "date": "Mon, 27 Mar 2023 13:49:59 GMT",
      "comments": {}
    },
    {
      "version": "0.2.8",
      "tag": "@dossierhq/core_v0.2.8",
      "date": "Mon, 27 Mar 2023 11:41:48 GMT",
      "comments": {
        "none": [
          {
            "comment": "Normalize/sort schema arrays"
          }
        ]
      }
    },
    {
      "version": "0.2.7",
      "tag": "@dossierhq/core_v0.2.7",
      "date": "Wed, 22 Mar 2023 21:32:23 GMT",
      "comments": {}
    },
    {
      "version": "0.2.6",
      "tag": "@dossierhq/core_v0.2.6",
      "date": "Wed, 22 Mar 2023 21:17:29 GMT",
      "comments": {}
    },
    {
      "version": "0.2.5",
      "tag": "@dossierhq/core_v0.2.5",
      "date": "Tue, 14 Mar 2023 17:47:19 GMT",
      "comments": {
        "none": [
          {
            "comment": "Break: Require camelCase for pattern names. Check type and field names. Export type AdminFieldSpecificationUpdate. Handle standalone carriage return (no line feed) in createRichTextTextAndLineBreakNodes(). Add createRichTextListNode()/createRichTextListItemNode()."
          }
        ]
      }
    },
    {
      "version": "0.2.4",
      "tag": "@dossierhq/core_v0.2.4",
      "date": "Tue, 07 Mar 2023 21:38:54 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add .client to Admin/PublishedExceptionClient."
          }
        ]
      }
    },
    {
      "version": "0.2.3",
      "tag": "@dossierhq/core_v0.2.3",
      "date": "Sat, 18 Feb 2023 11:05:21 GMT",
      "comments": {
        "none": [
          {
            "comment": "Break API: Remove isName from schema entity fields, add nameField to entity types. No break to database content. Introduce client.toExceptionClient() and Admin/PublishedExceptionClient."
          }
        ]
      }
    },
    {
      "version": "0.2.2",
      "tag": "@dossierhq/core_v0.2.2",
      "date": "Tue, 14 Feb 2023 18:18:06 GMT",
      "comments": {
        "none": [
          {
            "comment": "Rename validateTraverseNode() to validateTraverseNodeForSave(). Add validateTraverseNodeForPublish(), normalizeValueItem(), normalizeEntityFields(), groupValidationIssuesByTopLevelPath(). Enable typing for AdminCreate<AddType> in copyEntity(). Enable typing of unique index for clients."
          }
        ]
      }
    },
    {
      "version": "0.2.1",
      "tag": "@dossierhq/core_v0.2.1",
      "date": "Wed, 08 Feb 2023 18:24:39 GMT",
      "comments": {
        "none": [
          {
            "comment": "Highlight format in RichText"
          }
        ]
      }
    },
    {
      "version": "0.2.0",
      "tag": "@dossierhq/core_v0.2.0",
      "date": "Mon, 06 Feb 2023 21:59:52 GMT",
      "comments": {}
    },
    {
      "version": "0.1.53",
      "tag": "@dossierhq/core_v0.1.53",
      "date": "Mon, 06 Feb 2023 21:23:05 GMT",
      "comments": {
        "none": [
          {
            "comment": "Bump minor"
          }
        ]
      }
    },
    {
      "version": "0.1.52",
      "tag": "@dossierhq/core_v0.1.52",
      "date": "Tue, 31 Jan 2023 11:27:35 GMT",
      "comments": {
        "none": [
          {
            "comment": "Link node in RichText."
          }
        ]
      }
    },
    {
      "version": "0.1.51",
      "tag": "@dossierhq/core_v0.1.51",
      "date": "Sun, 29 Jan 2023 18:08:56 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support app client when converting json."
          }
        ]
      }
    },
    {
      "version": "0.1.50",
      "tag": "@dossierhq/core_v0.1.50",
      "date": "Sat, 28 Jan 2023 14:37:31 GMT",
      "comments": {}
    },
    {
      "version": "0.1.49",
      "tag": "@dossierhq/core_v0.1.49",
      "date": "Mon, 23 Jan 2023 09:38:38 GMT",
      "comments": {}
    },
    {
      "version": "0.1.48",
      "tag": "@dossierhq/core_v0.1.48",
      "date": "Sun, 22 Jan 2023 22:28:24 GMT",
      "comments": {}
    },
    {
      "version": "0.1.47",
      "tag": "@dossierhq/core_v0.1.47",
      "date": "Sun, 22 Jan 2023 22:01:52 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.46",
      "tag": "@dossierhq/core_v0.1.46",
      "date": "Tue, 17 Jan 2023 19:56:51 GMT",
      "comments": {}
    },
    {
      "version": "0.1.45",
      "tag": "@dossierhq/core_v0.1.45",
      "date": "Tue, 17 Jan 2023 17:55:40 GMT",
      "comments": {}
    },
    {
      "version": "0.1.44",
      "tag": "@dossierhq/core_v0.1.44",
      "date": "Tue, 17 Jan 2023 17:41:02 GMT",
      "comments": {}
    },
    {
      "version": "0.1.43",
      "tag": "@dossierhq/core_v0.1.43",
      "date": "Tue, 17 Jan 2023 10:34:29 GMT",
      "comments": {}
    },
    {
      "version": "0.1.42",
      "tag": "@dossierhq/core_v0.1.42",
      "date": "Tue, 17 Jan 2023 09:23:40 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add README"
          }
        ]
      }
    },
    {
      "version": "0.1.41",
      "tag": "@dossierhq/core_v0.1.41",
      "date": "Tue, 17 Jan 2023 08:54:04 GMT",
      "comments": {}
    },
    {
      "version": "0.1.40",
      "tag": "@dossierhq/core_v0.1.40",
      "date": "Mon, 16 Jan 2023 21:32:52 GMT",
      "comments": {}
    },
    {
      "version": "0.1.39",
      "tag": "@dossierhq/core_v0.1.39",
      "date": "Mon, 16 Jan 2023 21:07:04 GMT",
      "comments": {
        "none": [
          {
            "comment": "First release after @dossierhq/ move"
          }
        ]
      }
    },
    {
      "version": "0.1.38",
      "tag": "@jonasb/datadata-core_v0.1.38",
      "date": "Mon, 02 Jan 2023 22:59:38 GMT",
      "comments": {}
    },
    {
      "version": "0.1.37",
      "tag": "@jonasb/datadata-core_v0.1.37",
      "date": "Mon, 02 Jan 2023 17:39:42 GMT",
      "comments": {
        "none": [
          {
            "comment": "(break) Replace buildUrlWithUrlQuery, decodeUrlQueryStringifiedParam, stringifyUrlQueryParams with encodeObjectToURLSearchParams, decodeURLSearchParamsParam. Support typing of authKey. Improve typing of JSON. Enable checking if admin operation is modifying. Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.36",
      "tag": "@jonasb/datadata-core_v0.1.36",
      "date": "Mon, 12 Dec 2022 12:51:59 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support more rich text nodes. Make clients schema aware. Separate update schema types.  Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.35",
      "tag": "@jonasb/datadata-core_v0.1.35",
      "date": "Sat, 05 Nov 2022 21:33:57 GMT",
      "comments": {
        "none": [
          {
            "comment": "Sort schema. Add AdminSchema.createAndValidate(). Add isRichTextRootNode and isRichTextParagraphNode. Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.34",
      "tag": "@jonasb/datadata-core_v0.1.34",
      "date": "Tue, 25 Oct 2022 15:23:52 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support unique index"
          }
        ]
      }
    },
    {
      "version": "0.1.33",
      "tag": "@jonasb/datadata-core_v0.1.33",
      "date": "Mon, 17 Oct 2022 15:41:41 GMT",
      "comments": {
        "none": [
          {
            "comment": "Replace Temporal.Instant with Date. Support field matchPattern in schema and validator. Cache pattern regexps. Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.32",
      "tag": "@jonasb/datadata-core_v0.1.32",
      "date": "Fri, 30 Sep 2022 18:08:39 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support entity link node in RichText. Update dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.31",
      "tag": "@jonasb/datadata-core_v0.1.31",
      "date": "Mon, 19 Sep 2022 21:05:31 GMT",
      "comments": {}
    },
    {
      "version": "0.1.30",
      "tag": "@jonasb/datadata-core_v0.1.30",
      "date": "Mon, 19 Sep 2022 13:10:56 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support patterns in schema."
          }
        ]
      }
    },
    {
      "version": "0.1.29",
      "tag": "@jonasb/datadata-core_v0.1.29",
      "date": "Tue, 06 Sep 2022 09:47:25 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.28",
      "tag": "@jonasb/datadata-core_v0.1.28",
      "date": "Tue, 30 Aug 2022 21:10:47 GMT",
      "comments": {}
    },
    {
      "version": "0.1.27",
      "tag": "@jonasb/datadata-core_v0.1.27",
      "date": "Tue, 30 Aug 2022 14:05:13 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support adminOnly fields. Remove item visitor. Support published item traverser."
          }
        ]
      }
    },
    {
      "version": "0.1.26",
      "tag": "@jonasb/datadata-core_v0.1.26",
      "date": "Sun, 21 Aug 2022 22:26:51 GMT",
      "comments": {
        "none": [
          {
            "comment": "Make ValueItem more specific."
          }
        ]
      }
    },
    {
      "version": "0.1.25",
      "tag": "@jonasb/datadata-core_v0.1.25",
      "date": "Sun, 21 Aug 2022 12:14:33 GMT",
      "comments": {
        "none": [
          {
            "comment": "Switch rich text to Lexical."
          }
        ]
      }
    },
    {
      "version": "0.1.24",
      "tag": "@jonasb/datadata-core_v0.1.24",
      "date": "Mon, 13 Jun 2022 22:58:05 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support deeper typing of update and upsert."
          }
        ]
      }
    },
    {
      "version": "0.1.23",
      "tag": "@jonasb/datadata-core_v0.1.23",
      "date": "Mon, 13 Jun 2022 21:34:43 GMT",
      "comments": {}
    },
    {
      "version": "0.1.22",
      "tag": "@jonasb/datadata-core_v0.1.22",
      "date": "Mon, 13 Jun 2022 12:17:58 GMT",
      "comments": {
        "none": [
          {
            "comment": "Enable typing AdminEntity harder. Support multiline for String fields. Remove support for transparent json support of schema spec."
          }
        ]
      }
    },
    {
      "version": "0.1.21",
      "tag": "@jonasb/datadata-core_v0.1.21",
      "date": "Tue, 07 Jun 2022 21:00:46 GMT",
      "comments": {}
    },
    {
      "version": "0.1.20",
      "tag": "@jonasb/datadata-core_v0.1.20",
      "date": "Mon, 06 Jun 2022 22:23:03 GMT",
      "comments": {}
    },
    {
      "version": "0.1.19",
      "tag": "@jonasb/datadata-core_v0.1.19",
      "date": "Mon, 06 Jun 2022 15:26:55 GMT",
      "comments": {}
    },
    {
      "version": "0.1.18",
      "tag": "@jonasb/datadata-core_v0.1.18",
      "date": "Mon, 06 Jun 2022 14:27:59 GMT",
      "comments": {}
    },
    {
      "version": "0.1.17",
      "tag": "@jonasb/datadata-core_v0.1.17",
      "date": "Fri, 03 Jun 2022 19:47:15 GMT",
      "comments": {}
    },
    {
      "version": "0.1.16",
      "tag": "@jonasb/datadata-core_v0.1.16",
      "date": "Thu, 02 Jun 2022 22:03:43 GMT",
      "comments": {}
    },
    {
      "version": "0.1.15",
      "tag": "@jonasb/datadata-core_v0.1.15",
      "date": "Thu, 02 Jun 2022 21:27:42 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies"
          }
        ]
      }
    },
    {
      "version": "0.1.14",
      "tag": "@jonasb/datadata-core_v0.1.14",
      "date": "Mon, 30 May 2022 18:59:49 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add assertOkResult, assertErrorResultType"
          }
        ]
      }
    },
    {
      "version": "0.1.13",
      "tag": "@jonasb/datadata-core_v0.1.13",
      "date": "Mon, 30 May 2022 15:28:33 GMT",
      "comments": {
        "none": [
          {
            "comment": "Replace typescript enums with const {}"
          }
        ]
      }
    },
    {
      "version": "0.1.12",
      "tag": "@jonasb/datadata-core_v0.1.12",
      "date": "Mon, 30 May 2022 11:14:25 GMT",
      "comments": {
        "none": [
          {
            "comment": "Output ESM and CJS."
          }
        ]
      }
    },
    {
      "version": "0.1.11",
      "tag": "@jonasb/datadata-core_v0.1.11",
      "date": "Tue, 24 May 2022 15:02:06 GMT",
      "comments": {}
    },
    {
      "version": "0.1.10",
      "tag": "@jonasb/datadata-core_v0.1.10",
      "date": "Tue, 17 May 2022 22:02:29 GMT",
      "comments": {}
    },
    {
      "version": "0.1.9",
      "tag": "@jonasb/datadata-core_v0.1.9",
      "date": "Tue, 17 May 2022 21:38:25 GMT",
      "comments": {}
    },
    {
      "version": "0.1.8",
      "tag": "@jonasb/datadata-core_v0.1.8",
      "date": "Tue, 17 May 2022 18:54:10 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support fieldItem in traverseAdminItem. Add Ok/ErrorFromResult."
          }
        ]
      }
    },
    {
      "version": "0.1.7",
      "tag": "@jonasb/datadata-core_v0.1.7",
      "date": "Tue, 22 Mar 2022 23:23:55 GMT",
      "comments": {}
    },
    {
      "version": "0.1.6",
      "tag": "@jonasb/datadata-core_v0.1.6",
      "date": "Tue, 22 Mar 2022 21:24:28 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add PagingInfo type"
          }
        ]
      }
    },
    {
      "version": "0.1.5",
      "tag": "@jonasb/datadata-core_v0.1.5",
      "date": "Sat, 12 Mar 2022 09:19:52 GMT",
      "comments": {}
    },
    {
      "version": "0.1.4",
      "tag": "@jonasb/datadata-core_v0.1.4",
      "date": "Mon, 07 Mar 2022 22:35:33 GMT",
      "comments": {}
    },
    {
      "version": "0.1.3",
      "tag": "@jonasb/datadata-core_v0.1.3",
      "date": "Sun, 06 Mar 2022 19:47:26 GMT",
      "comments": {}
    },
    {
      "version": "0.1.2",
      "tag": "@jonasb/datadata-core_v0.1.2",
      "date": "Mon, 28 Feb 2022 22:25:07 GMT",
      "comments": {
        "none": [
          {
            "comment": "Support entity sampling. Support advisory locks"
          }
        ]
      }
    },
    {
      "version": "0.1.1",
      "tag": "@jonasb/datadata-core_v0.1.1",
      "date": "Sun, 13 Feb 2022 21:58:15 GMT",
      "comments": {
        "none": [
          {
            "comment": "Add OkResult.map() and ErrorType.isErrorType()"
          },
          {
            "comment": "Configure importHelpers for ts, and add tslib dependency"
          },
          {
            "comment": "Export transpiled JavaScript"
          },
          {
            "comment": "Fix export configuration"
          },
          {
            "comment": "Allow string in schema field specification"
          },
          {
            "comment": "Enable easier JSON serialization of adminClient"
          },
          {
            "comment": "Support boolean type"
          },
          {
            "comment": "upsertEntity() support in AdminClient"
          },
          {
            "comment": "Make AdminEntityUpsert.fields required"
          },
          {
            "comment": "Normalize field values. Effect on create/update entity"
          },
          {
            "comment": "Add isEntityNameAsRequested()"
          },
          {
            "comment": "Normalize value items with missing fields'"
          },
          {
            "comment": "Add Conflict as error to createEntity(). Upgrade dependencies"
          },
          {
            "comment": "createdAt/updatedAt sort order"
          },
          {
            "comment": "Use Temporal, createdAt, updatedAt, random tweaks"
          },
          {
            "comment": "Minor tweaks"
          },
          {
            "comment": "Support context provider for published client"
          },
          {
            "comment": "Convert instants for searchAdminEntities() for JSON, require logger for client context"
          },
          {
            "comment": "Add NoOpLogger, createConsoleLogger"
          },
          {
            "comment": "Add url query utils"
          },
          {
            "comment": "Add searchEntities and getTotalCount to PublishedClient"
          },
          {
            "comment": "Add JSON support to PublishedClient"
          },
          {
            "comment": "Add getSchemaSpecification() to PublishedClient. Fix JSON conversion of getEntities()."
          },
          {
            "comment": "Upgrade dependencies. Support adminOnly."
          },
          {
            "comment": "Status filter"
          },
          {
            "comment": "Support ItemValuePath"
          },
          {
            "comment": "Support required fields. Add item traverser. "
          },
          {
            "comment": "Support reverse order"
          },
          {
            "comment": "Support authKeys"
          },
          {
            "comment": "Add createdAt to EntityInfo"
          },
          {
            "comment": "Upgrade depencencies. Renames by adding Published prefix. Add {publish:true}."
          },
          {
            "comment": "Remove authKeys from reference to entity (they always use the actual authkey of the entity). Support sampleEntities(). Add optional id to RichTextBlock."
          },
          {
            "comment": "Support interface for buildUrlWithUrlQuery"
          }
        ]
      }
    },
    {
      "version": "0.1.0",
      "tag": "@jonasb/datadata-core_v0.1.0",
      "date": "Mon, 12 Jul 2021 17:25:50 GMT",
      "comments": {
        "none": [
          {
            "comment": "Initial release"
          }
        ]
      }
    }
  ]
}
