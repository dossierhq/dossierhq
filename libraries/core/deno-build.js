#!/usr/bin/env node
const fs = require('fs/promises');
const path = require('path');
const ts = require('typescript');

function rewrittenRelativeImport(moduleSpecifier) {
  if (moduleSpecifier.endsWith('.')) moduleSpecifier += '/';
  if (moduleSpecifier.endsWith('/')) moduleSpecifier += 'index.js';
  if (!moduleSpecifier.endsWith('.js')) moduleSpecifier += '.js';
  return moduleSpecifier;
}

function rewriteNpmImport(package, _version) {
  return package;
}

function rewriteImportExportModuleSpecifier(moduleSpecifier, dependencies) {
  const isRelative = moduleSpecifier.startsWith('.');
  let rewritten = moduleSpecifier;
  if (isRelative) {
    rewritten = rewrittenRelativeImport(rewritten);
  } else {
    const version = dependencies[moduleSpecifier];
    if (version) {
      rewritten = rewriteNpmImport(moduleSpecifier, version);
    }
  }
  return rewritten;
}

function rewriteImports(context) {
  const packageJson = require('./package.json');
  const { dependencies } = packageJson;

  return (sourceFile) => {
    function visitor(node) {
      if (ts.isImportDeclaration(node)) {
        const moduleSpecifier = node.moduleSpecifier.text;
        const rewritten = rewriteImportExportModuleSpecifier(moduleSpecifier, dependencies);
        if (moduleSpecifier !== rewritten) {
          return context.factory.updateImportDeclaration(
            node,
            node.decorators,
            node.modifiers,
            node.importClause,
            context.factory.createStringLiteral(rewritten)
          );
        }
      } else if (ts.isExportDeclaration(node)) {
        const moduleSpecifier = node.moduleSpecifier?.text;
        if (moduleSpecifier) {
          const rewritten = rewriteImportExportModuleSpecifier(moduleSpecifier, dependencies);
          if (moduleSpecifier !== rewritten) {
            return context.factory.updateExportDeclaration(
              node,
              node.decorators,
              node.modifiers,
              node.isTypeOnly,
              node.exportClause,
              context.factory.createStringLiteral(rewritten)
            );
          }
        }
      }

      return ts.visitEachChild(node, visitor, context);
    }
    return ts.visitNode(sourceFile, visitor);
  };
}

async function buildTypescript(basePath) {
  const configFileName = ts.findConfigFile(basePath, ts.sys.fileExists, 'tsconfig.deno.json');
  const configFile = ts.readConfigFile(configFileName, ts.sys.readFile);
  if (configFile.error) {
    handleDiagnosticError(configFile.error);
  }
  configFile.compilerOptions = { ...configFile.compilerOptions };
  const commandLine = ts.parseJsonConfigFileContent(configFile.config, ts.sys, basePath);

  if (commandLine.errors.length > 0) {
    handleDiagnosticErrors(commandLine.errors);
  }
  const compilerOptions = commandLine.options;
  const program = ts.createProgram({
    rootNames: commandLine.fileNames,
    options: compilerOptions,
  });
  const emitResult = program.emit(undefined, undefined, undefined, undefined, {
    after: [rewriteImports],
    afterDeclarations: [rewriteImports],
  });
  if (emitResult.errors?.length > 0) {
    handleDiagnosticErrors(emitResult.errors);
  }
  const preEmitDiagnostics = ts.getPreEmitDiagnostics(program);
  if (preEmitDiagnostics.length > 0) {
    handleDiagnosticErrors(preEmitDiagnostics);
  }

  const files = await findFilesRecursive(compilerOptions.outDir, (filename) =>
    filename.endsWith('.js')
  );
  for (const outputJsFile of files) {
    let contents = await fs.readFile(outputJsFile, { encoding: 'utf-8' });
    const basename = path.basename(outputJsFile);
    contents = `/// <reference types="./${basename.slice(0, -3)}.d.ts" />\n${contents}`;
    await fs.writeFile(outputJsFile, contents);
  }
}

async function findFilesRecursive(directory, filter) {
  const entries = (await fs.readdir(directory)).map((f) => `${directory}/${f}`);
  const result = entries.filter(filter);
  for (const entry of entries) {
    const stat = await fs.stat(entry);
    if (stat.isDirectory()) {
      result.push(...(await findFilesRecursive(entry, filter)));
    }
  }
  return result;
}

function handleDiagnosticError(error) {
  console.log(
    ts.formatDiagnostic(error, {
      getCanonicalFileName: (path) => path,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      getNewLine: () => ts.sys.newLine,
    })
  );
  throw new Error();
}

function handleDiagnosticErrors(errors) {
  console.log(
    ts.formatDiagnosticsWithColorAndContext(errors, {
      getCanonicalFileName: (path) => path,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      getNewLine: () => ts.sys.newLine,
    })
  );
  throw new Error();
}

buildTypescript(process.cwd()).catch((error) => {
  console.log(error);
  process.exit(1);
});
