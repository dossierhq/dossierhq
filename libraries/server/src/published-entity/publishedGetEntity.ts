import type {
  EntityReferenceWithAuthKeys,
  ErrorType,
  PromiseResult,
  PublishedEntity,
  PublishedSchema,
} from '@jonasb/datadata-core';
import { notOk, ok } from '@jonasb/datadata-core';
import type { AuthorizationAdapter, DatabaseAdapter, SessionContext } from '..';
import { authVerifyAuthorizationKey } from '../Auth';
import * as Db from '../Database';
import type { EntitiesTable, EntityVersionsTable } from '../DatabaseTables';
import { decodePublishedEntity } from '../EntityCodec';

export async function publishedGetEntity(
  schema: PublishedSchema,
  authorizationAdapter: AuthorizationAdapter,
  databaseAdapter: DatabaseAdapter,
  context: SessionContext,
  reference: EntityReferenceWithAuthKeys
): PromiseResult<
  PublishedEntity,
  ErrorType.BadRequest | ErrorType.NotFound | ErrorType.NotAuthorized | ErrorType.Generic
> {
  const entityMain = await Db.queryNoneOrOne<
    Pick<
      EntitiesTable,
      'uuid' | 'type' | 'name' | 'auth_key' | 'resolved_auth_key' | 'created_at'
    > &
      Pick<EntityVersionsTable, 'data'>
  >(
    databaseAdapter,
    context,
    `SELECT e.uuid, e.type, e.name, e,auth_key, e.resolved_auth_key, e.created_at, ev.data
      FROM entities e, entity_versions ev
      WHERE e.uuid = $1
      AND e.published_entity_versions_id = ev.id`,
    [reference.id]
  );
  if (!entityMain) {
    return notOk.NotFound('No such entity');
  }

  const authResult = await authVerifyAuthorizationKey(
    authorizationAdapter,
    context,
    reference?.authKeys,
    { authKey: entityMain.auth_key, resolvedAuthKey: entityMain.resolved_auth_key }
  );
  if (authResult.isError()) {
    return authResult;
  }

  const entity = decodePublishedEntity(schema, entityMain);

  return ok(entity);
}
