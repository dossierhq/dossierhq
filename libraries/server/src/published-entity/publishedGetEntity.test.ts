import { ok } from '@dossierhq/core';
import { describe, expect, test } from 'vitest';
import {
  ENCODE_VERSION_AS_IS,
  ENCODE_VERSION_LEGACY,
} from '../shared-entity/migrateDecodeAndNormalizeEntityFields.js';
import {
  createMockAuthorizationAdapter,
  createMockDatabaseAdapter,
  createMockSessionContext,
  getDatabaseAdapterMockedCallsWithoutContextAndUnordered,
  type MockDatabaseAdapter,
} from '../test/AdditionalTestUtils.js';
import { adminTestSchema } from '../test/TestSchema.js';
import { publishedGetEntity } from './publishedGetEntity.js';

function mockPublishedEntityGetOneCall(
  databaseAdapter: MockDatabaseAdapter,
  {
    type,
    encodeVersion,
    fields,
  }: { type: string; encodeVersion?: number; fields: Record<string, unknown> },
) {
  const date = new Date('2023-08-09T08:22:35.810Z');

  databaseAdapter.publishedEntityGetOne.mockReturnValueOnce(
    Promise.resolve(
      ok({
        id: '123',
        name: 'Name',
        type,
        authKey: 'none',
        resolvedAuthKey: 'none',
        createdAt: date,
        validPublished: true,
        entityFields: {
          schemaVersion: 1,
          encodeVersion: encodeVersion ?? ENCODE_VERSION_AS_IS,
          fields,
        },
      }),
    ),
  );
}

describe('publishedGetEntity', () => {
  test('Minimal', async () => {
    const databaseAdapter = createMockDatabaseAdapter();
    const authorizationAdapter = createMockAuthorizationAdapter();
    const context = createMockSessionContext({ databaseAdapter });

    mockPublishedEntityGetOneCall(databaseAdapter, {
      type: 'TitleOnly',
      fields: { title: 'Title' },
    });

    const result = await publishedGetEntity(
      adminTestSchema,
      authorizationAdapter,
      databaseAdapter,
      context,
      { id: '123' },
    );

    expect(result.valueOrThrow()).toMatchSnapshot();
    expect(
      getDatabaseAdapterMockedCallsWithoutContextAndUnordered(databaseAdapter),
    ).toMatchSnapshot();
  });

  test('Legacy codec', async () => {
    const databaseAdapter = createMockDatabaseAdapter();
    const authorizationAdapter = createMockAuthorizationAdapter();
    const context = createMockSessionContext({ databaseAdapter });

    mockPublishedEntityGetOneCall(databaseAdapter, {
      type: 'ComponentsEntity',
      encodeVersion: ENCODE_VERSION_LEGACY,
      fields: {
        // References were stored as just the id
        normal: { type: 'EntitiesComponent', normal: '12345', list: ['23456', '34567'] },
        list: [
          // Locations were stored as [lat, lng]
          {
            type: 'LocationsComponent',
            normal: [1, 2],
            list: [
              [3, 4],
              [5, 6],
            ],
          },
        ],
      },
    });

    const result = await publishedGetEntity(
      adminTestSchema,
      authorizationAdapter,
      databaseAdapter,
      context,
      { id: '123' },
    );

    expect(result.valueOrThrow()).toMatchSnapshot();
  });
});
