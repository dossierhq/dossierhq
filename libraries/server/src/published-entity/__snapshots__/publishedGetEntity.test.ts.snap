// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`publishedGetEntity > Legacy codec 1`] = `
{
  "fields": {
    "list": [
      {
        "list": [
          {
            "lat": 3,
            "lng": 4,
          },
          {
            "lat": 5,
            "lng": 6,
          },
        ],
        "normal": {
          "lat": 1,
          "lng": 2,
        },
        "type": "LocationsValueItem",
      },
    ],
    "normal": {
      "list": [
        {
          "id": "23456",
        },
        {
          "id": "34567",
        },
      ],
      "normal": {
        "id": "12345",
      },
      "type": "EntitiesValueItem",
    },
  },
  "id": "123",
  "info": {
    "authKey": "none",
    "createdAt": 2023-08-09T08:22:35.810Z,
    "name": "Name",
    "type": "ValueItemsEntity",
    "valid": true,
  },
}
`;

exports[`publishedGetEntity > Minimal 1`] = `
{
  "fields": {
    "title": "Title",
  },
  "id": "123",
  "info": {
    "authKey": "none",
    "createdAt": 2023-08-09T08:22:35.810Z,
    "name": "Name",
    "type": "TitleOnly",
    "valid": true,
  },
}
`;

exports[`publishedGetEntity > Minimal 2`] = `
[
  [
    "publishedEntityGetOne",
    {
      "id": "123",
    },
  ],
]
`;
